Add handler for creating WebSocket connection

igorveselinovic opened the pull request:
Resolves #7 

The endpoint `/patches/v1/connect/{conversation_id}` creates a WebSocket (WS) connection between a client and `patches`.

Once the connection is upgraded to a WS connection, `patches` starts a goroutine to start handling messages in the WS. Before reading patches from the WS and writing patches to the WS, `patches` expects the first message to be just the token (should this maybe be sent as a JSON object with one `token` field?). `patches` then makes an HTTP request to `heimdall` to validate the token.

If the token is valid, `patches` then starts a `Client` with one goroutine for reading from the WS and one goroutine for writing to the WS. If this is the first client associated with a conversation, then `patches` creates a `Conversation` and makes an HTTP request to `ether` to get the conversation content. For now, this is just stored as a string in memory, and I'm not sure how else to deal with this right now.

The `Broker` struct contains the methods that are responsible for creation/deletion of `Client` structs and `Conversation` structs.

igorveselinovic committed:
Fix formatting of Go code

igorveselinovic committed:
Remove tmp folder

igorveselinovic committed:
Add initial janky WebSockets logic

Not fully thread-safe and doesn't clean up properly yet.

igorveselinovic committed:
Make WebSocket handling thread-safe

There is a struct for managing a single conversation, a struct for
managing a single WebSocket connection, and a struct for connecting the
two together in a thread-safe way.

igorveselinovic committed:
Change where goroutine is used

igorveselinovic committed:
Add code for getting convo content from Ether

Getting content after verifying the token, and getting the user ID.
Getting the content also inherently acts as a check for whether the user
is in the conversation.

igorveselinovic committed:
Add .gitignore

igorveselinovic committed:
Merge branch 'master' into igor/websockets

igorveselinovic committed:
Remove unused constant

igorveselinovic committed:
Add code comments for new stuff

igorveselinovic commented:
While making this PR, I realized that after validating the token, but before registering a new `Client`, `patches` needs to make a call to `/ether/v1/conversations/{conversation_id}/users/{user_id}` to validate that the user who made the WS connection is actually in the conversation.

So I'll fix that.

igorveselinovic committed:
Add check for conversation membership

Before registering a Client, make sure that the user is actually in the
conversation. Copied over the UserConversationMapping struct from ether.

Thao-Tran commented on websockets/broker.go:
```diff
@@ -0,0 +1,239 @@
+)
+
+var heimdallHost = os.Getenv("HEIMDALL_HOST")
+var etherHost = os.Getenv("ETHER_HOST")
```
```suggestion
var (
	heimdallHost = os.Getenv("HEIMDALL_HOST")
	etherHost = os.Getenv("ETHER_HOST")
)
```

Thao-Tran commented on websockets/client.go:
```diff
@@ -0,0 +1,65 @@
+// sent into the Client's channel.
+func (c *Client) write() {
+	for {
+		message, ok := <-c.send
```
```suggestion
	for message, ok := range <-c.send
```

igorveselinovic replied:
You mean this ?
```suggestion
	for message, ok := <-c.send {
```

Thao-Tran replied:
Oh no sorry I meant this.
```suggestion
	for message, ok := range c.send
```

igorveselinovic replied:
You mean this ?
```suggestion
	for message, ok := range c.send {
```

Thao-Tran replied:
Haha yeah like the second one.

Thao-Tran commented on websockets/conversation.go:
```diff
@@ -0,0 +1,85 @@
+			c.clients[client] = true
+			err := client.conn.WriteMessage(gorillaws.TextMessage, []byte(c.doc))
+			if err != nil {
+				return
```
Should probably log the error. Also would we want to stop the goroutine entirely in this case or just `continue`?

igorveselinovic replied:
Yeah, should log the error and `continue`. Maybe also shut down the client?

Thao-Tran replied:
Yeah true, it'd probably be a good idea to shut down the client too.

Thao-Tran commented on websockets/conversation.go:
```diff
@@ -0,0 +1,85 @@
+
+			patches, err := dmp.PatchFromText(string(message.content))
+			if err != nil {
+				return
```
Same as previous comment.

Thao-Tran commented on websockets/broker.go:
```diff
@@ -0,0 +1,239 @@
+
+// Broker is the single entrypoint for registering and unregistering all
+// client connections.
+type Broker struct {
```
Maybe you can have a `Run` method for `Broker` that runs in a goroutine and just reads from `register` and `unregister` channels so that you don't have to use locking and whatnot. Then, in `StartClient`, you would just send the member and connection to the broker's `register` channel.

igorveselinovic replied:
As we talked about, it makes sense, but just for the sake I've time I'll leave it as-is for now.

Thao-Tran commented on websockets/conversation.go:
```diff
@@ -0,0 +1,85 @@
+	conversationID int64
+	doc            string
+	clients        map[*Client]bool
+	broker         *Broker
```
Does a Conversation need to store a Broker?

igorveselinovic replied:
It does not. Good catch.

Thao-Tran commented on websockets/client.go:
```diff
@@ -0,0 +1,65 @@
+type Client struct {
+	userID       int64
+	conn         *gorillaws.Conn
+	conversation *Conversation
```
Like we discussed before, it might be better to just have a `broadcast` channel and a `conversationID` int64 to replace the `conversation` field.

igorveselinovic replied:
Will do.

Thao-Tran commented on websockets/conversation.go:
```diff
@@ -0,0 +1,85 @@
+			if err != nil {
+				return
+			}
+			c.doc, _ = dmp.PatchApply(patches, c.doc)
```
Did you purposefully ignore the error since the error handling will be a whole other issue?

igorveselinovic replied:
For handling this error, should we maybe just log it and close the WebSocket that sent the "bad" patch for now?

Thao-Tran replied:
Yeah that sounds good.

Thao-Tran commented on handlers/handlers.go:
```diff
@@ -50,3 +72,23 @@ func (env *Env) GetPatchesHandler(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	go env.WSBroker.StartClient(conversationID, c)
```
Do we not need to respond with any errors if the client isn't able to start?

igorveselinovic replied:
Correct me if I'm wrong, but at this point, I think the connection has already been upgraded to a WebSocket connection, so I don't know if you can respond.

Thao-Tran replied:
Okay yeah I actually had no idea so please correct me if I'm wrong haha

igorveselinovic replied:
Yeah once it upgrades to the WebSocket protocol, it should just use that to communicate, as far as I understand.

Thao-Tran replied:
Okay cool sounds good. Carry on then.

igorveselinovic committed:
Update websockets/broker.go

Co-Authored-By: Thao-Tran Le-Phuong <Thao-Tran@users.noreply.github.com>

igorveselinovic committed:
Update websockets/client.go

igorveselinovic committed:
Fix ranging through the client.send channel

The loop just exits once the channel is closed (i.e., when ok would be
set to false), so that needed to be fixed.

igorveselinovic committed:
Remove broker & add error handling to Conversation

For now, just close the relevant Client channel when something goes
wrong, log the error, and continue the goroutine.

igorveselinovic committed:
Update Client to store convo ID and broadcast chan

Doing this instead of storing the whole Conversation struct, to make the
two less tightly couple.

igorveselinovic committed:
Update Broker to work with other struct changes

igorveselinovic committed:
Add conversationID to NewClient function

Also, add some error logging.
